extensions/applicationinsights-clickanalytics-js/src/Interfaces/Datamodel.ts

```diff
@@ -0,0 +1,283 @@
+     */
+    urlCollectQuery?: boolean;
+    /**
+     * Validator to use for the data-bhvr value
+     */
+    behaviorValidator?: (value: string | number) => string | number;
+}
+
+/**
+ * Custom Data Tags Configuration
+ */
+export interface ICustomDataTags {
+    /**
+     * When a particular element is not tagged with content name prefix or content name prefix is not provided by user, this flag is used to collect standard HTML attribute for contentName.
+     */
+    useDefaultContentName?: boolean;
+    /**
+     * Automatic capture content name and value of elements which are tagged with provided prefix
+     */
+    customDataPrefix?: string;
+    /**
+     * Click Analytics supports a JSON blob content meta data tagging instead of individual data-* attributes. The default attribute is data-ai-blob. This property allows for changing that attribute name.
+     */
+    aiBlobAttributeTag?: string;
+    /**
+     * Automatic capture metadata name and content with provided prefix
+     */
+    metaDataPrefix?: string;
+    /**
+     * Automatic capture all metadata names and content. Default is false. If enabled this will override provided metaTagPrefix.
+     */
+    captureAllMetaDataContent?: boolean;
+    /**
+     * Stop traversing up the DOM to capture content name and value of elements when encountered with this tag
+     */
+    parentDataTag?: string;
+    /**
+     * Custom attribute Tag to not track telemetry data
+     */
+    donotTrackDataTag?: string
+}
+
+/**
+ * Core data configuration
+ */
+export interface ICoreData {
+    /**
+     * document.referrer is the default. This is used to override the default value.
+     */
+    referrerUri?: string;
+    /**
+     * window.location.href is the default. This is used to override the default value.
+     */
+    requestUri?: string;
+    /**
+     * Default page name is derived from the url. This is used to override the default.
+     */
+    pageName?: string;
+    /**
+     * PageType is captured from a meta tag named awa-pageType. This is used to override the value or bypass defining meta tags.
+     */
+    pageType?: string;
+}
+
+
+/**
+ * Value Callbacks configuration
+ */
+export interface IValueCallback {
+
+    /**
+     * Function to override the default pageName capturing behavior.
+     */
+    pageName?: () => string;
+    /**
+     * A callback function to augument the default pageTags collected during pageAction event.
+     */
+    pageActionPageTags?: (element?: Element) => IPageTags;
+    /**
+     * A callback function to populate customized contentName.
+     */
+    contentName?: (element?: any, useDefaultContentName?: boolean) => string;
+}
+
+
+/**
+ * PageTags format
+ */
+export interface IPageTags {
+   /**
+    * Meta data tags
+    */
+    metaTags?: { [name: string]: string };
+   /**
+    * Any other page tag
+    */
+    [name: string]: string | number | boolean | string[] | number[] | boolean[] | object;
```

Comment:
FYI - Just did an internal fix for this line to include "undefined" as an option.

Why? When users of the NPM (consuming the TS files) are using "strict" mode TypeScript complains that the optional fields like "metaTags?" is a string | undefined and it can't be assigned to this line. This issue affects all interfaces / classes that have this "[name: string]: xxx" at the top level with optional properties.